src/06_ffi.md

```diff
@@ -0,0 +1,395 @@
+}
+#[repr(C, packed)]
+struct PackedData {
+    a: u32,
+    b: u16,
+    c: u64,
+}
+```
+
+are compatible with the following C types:
+
+```c
+struct Data {
+    uint32_t a;
+    uint16_t b;
+    uint64_t c;
+};
+__attribute__((packed))
+struct PackedData {
+    uint32_t a;
+    uint16_t b;
+    uint64_t c;
+}
+```
+
+> ### Rule {{#check FFI-CTYPE | Use only C-compatible types in FFI}}
+>
+> In a secure Rust development, only C-compatible types must be used as
+> parameter or return type of imported or exported functions and as types of
+> imported or exported global variables.
+
+The following types are considered C-compatible:
+
+- integral or floating point primitive types,
+- `repr(C)`-annotated `struct`,
+- `repr(C)` or `repr(Int)`-annotated `enum` with only fieldless variants (where
+  `Int` is an integral primitive type),
+- pointers.
+
+The following types are not C-compatible:
+
+- Dynamically sized types,
+- Trait objects,
+- Enums with fields,
+- Tuples (but `repr(C)` tuple structures are OK).
+
+Some types are compatibles with some caveats:
+
+- Zero-sized types, which is really zero sized (which is let unspecified in C
+  and contradicts the C++ specification),
+- `repr(C)`, `repr(C, Int)`, or `repr(Int)`-annotated enum with fields
+  ([RFC 2195]).
+
+References:
+
+- [RFC 2195 (Really tagged unions)][RFC 2195]
+- [Rust Reference: Type Layout]
+- [Rust Book: Unsafe Rust]
+
+[RFC 2195]: https://github.com/rust-lang/rfcs/blob/master/text/2195-really-tagged-unions.md
+[Rust Reference: Type Layout]: https://doc.rust-lang.org/reference/type-layout.html
+[Rust Book: Unsafe Rust]: https://doc.rust-lang.org/book/ch19-01-unsafe-rust.html
+
+## Pointers and references
+
+Although there are allowed by the Rust compiler, the use of Rust references in
+FFI should be avoided. It is particularly true when binding to and from C,
+because C has no references (in the sense of valid pointers):
+
+> ### Rule {{#check FFI-NOREF | Do not use unchecked references in FFI}}
+>
+> In a secure Rust development, there must be no Rust reference types when
+> interfacing with C, in particular, in:
+>
+> - the prototype of an imported or exported function,
+> - the type of an imported or exported global variables,
+>
+> directly or indirectly (in a non-opaque subtype).
+
+In case of binding to or from C++, it is possible to use Rust references on one
+side, and C++ references on the other. However, the C++ code should be checked
+against pointer/reference confusion.
+
+> ### Rule {{#check FFI-CKPTR | Check foreign pointers}}
+>
+> In a secure Rust development, any Rust code that dereferences a foreign
+> pointer must check their validity beforehand.
+> In particular, pointers must be checked to be non-null before any use.
+
+Stronger approaches such as _tagged pointers_ are possible if the pointed value
+is only manipulated from the Rust side of an FFI boundary.
+
+> ### Note
+>
+> It is possible to use the type `Option<&T>` (resp. `Option<&mut T>`) instead
```

Comment:
This optimization is (one of the few things) guaranteed, although `T` needs to be `Sized`.

  - An additional note could be added against never explicitely dereferencing raw pointers, but always using `.as_ref()` and `.as_mut` like in the examples below (for pointers pointing to initialized data)